pallene/coder.lua

```diff
@@ -1398,8 +1398,12 @@ gen_cmd["CallStatic"] = function(self, cmd, func)
     local top = self:stack_top_at(func, cmd)
 
     local parts = {}
-    table.insert(parts, self:call_pallene_function(dsts, #func.captured_vars >= 1,
-        cmd.f_id, top, xs))
+
+    -- CallStatic is used in pallene-pallene calls. Any calls to higher order functions or
+    -- closures is always done using CallDyn at the moment, meaning CallStatic will never call
+    -- a closure with upvalues.
```

Comment:
I'm not sure that this assumption will still hold after next week. Once we split the `G`, the global variables will become upvalues.

Instead of the `call_pallene_function` receiving a `has_upvalue`, it should receive a string with a `C` expression for the `CClosure*`. In the Lua entry point that is the `func`, but in the Pallene entry point it will be something else.  (Currently, the `"func"` in the `call_pallene_function` assumes that we're inside a Lua entry point, which doesn't smell right)

I think that the `call_pallene_function` should be able to know how many upvalues the function has, based on the `f_id`. We might be able to do the NULL optimization inside the `call_pallene_function`, instead of in the places that call it.

Reply:
At some point, the call static functions certainly will start having upvalues too. Either after we split the `G`, or after we implement #414, whichever comes first.

At the very least, we need to rewrite this comment. While this might be true today, it certalny won't be true anymore in the near future. However, I feel that maybe we should bite the bullet and try to allow upvalues here as well.

The impression I'm having here is that the only reason we added the assert, is to avoid having to come up with a reference to the cclosure. However, getting that cclosure is going to be a pre-requesite for #413. I suppose we might as well do it now...